### ISSUE TITLE
[question]Some questions about configuring roles and claims

### ISSUE THREAD

**Issue by qq1176914912** [simpleidserver/SimpleIdServer#632]
Sorry to bother you, I need to consult with some questions.
1. When configuring 'clients', there is an option for' roles'. After I configure it, the user client logs in to access_ There is no such character in the token, so what is the purpose of this option and how should it be used?
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/80dcb594-89b9-404e-9065-eb66f631c64a)
2. When configuring 'Users', there are' Claims' and 'Groups' options. After I configure them, I use this account to log in to my client and access the API. The API will retrieve User. Claims, but the obtained Claims do not have the values I set. How do I use these two options?
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/eae42ba0-f7a5-44e4-b77b-0fac7d300b95)

So what do I need to do to get the roles or groups?

**Comment by simpleidserver** [comment]
**How to Assign Roles to a User** ?

When a role is added to a client, a new scope named `<client_id>\<role_name>` is added and assigned to the client.

Once the role is defined, you can create a group and assign roles to it.
Follow these instructions:

1. On the Groups screen, click on the `Add group` button, fill in the form, and click on the `Submit` button.
2. Click on the new group and navigate to the `Roles` tab.
3. Click on the `Add Role` button, select your role, and click on the Save button.

Now, there is a group configured with one role; you can assign it to the User.

1. On the `Users` screen, select the User and navigate to the `Groups` tab.
2. Click on the `Assign groups` button and assign one or more groups.
3. Navigate to the `Claims` tab, click on the `Resolve roles` checkbox, and check if the roles are displayed in the table.

**How to Add Custom Claims to the id_token** ?

In SimpleIdServer, you can define an `Identity Scope` with one or more `Mapping Rules`.
These rules are used to retrieve user information and convert it into a list of claims that will be included in the Identity Token.

To return your custom `testclaims` into the Identity Token, you must create a SCOPE with one mapping rule:

1. On the `Scopes` screen, click on the `Add scope` button.
2. Select `Identity Value` and click on next.
3. Fill in the form like this and click on the Save button to confirm the creation.

| Parameter   | Value  |
| ----------- | ------ |
| name        | test   |
| description | test   |
| protocol    | openid |
| Is Exposed   | false  |

4. Navigate to the new scope, click on the `Mappers` tab, and click on the `Add mapper` button.
5. Select `Attribute` and click on next.
6. Fill in the form like this and click on the Save button to confirm the creation.

| Parameter        | Value      |
| ---------------- | ---------- |
| name             | test       |
| token claim name | test       |
| claim json type  | string     |
| user attribute   | testclaims |

Now the scope is configured; you can pass it into the authorization request. The value of the`testclaims` claim will be passed into the identity token.

**Comment by qq1176914912** [comment]
> **How to Assign Roles to a User** ?
> 
> When a role is added to a client, a new scope named `<client_id>\<role_name>` is added and assigned to the client.
> 
> Once the role is defined, you can create a group and assign roles to it. Follow these instructions:
> 
> 1. On the Groups screen, click on the `Add group` button, fill in the form, and click on the `Submit` button.
> 2. Click on the new group and navigate to the `Roles` tab.
> 3. Click on the `Add Role` button, select your role, and click on the Save button.
> 
> Now, there is a group configured with one role; you can assign it to the User.
> 
> 1. On the `Users` screen, select the User and navigate to the `Groups` tab.
> 2. Click on the `Assign groups` button and assign one or more groups.
> 3. Navigate to the `Claims` tab, click on the `Resolve roles` checkbox, and check if the roles are displayed in the table.
> 
> **How to Add Custom Claims to the id_token** ?
> 
> In SimpleIdServer, you can define an `Identity Scope` with one or more `Mapping Rules`. These rules are used to retrieve user information and convert it into a list of claims that will be included in the Identity Token.
> 
> To return your custom `testclaims` into the Identity Token, you must create a SCOPE with one mapping rule:
> 
> 1. On the `Scopes` screen, click on the `Add scope` button.
> 2. Select `Identity Value` and click on next.
> 3. Fill in the form like this and click on the Save button to confirm the creation.
> 
> Parameter	Value
> name	test
> description	test
> protocol	openid
> Is Exposed	false
> 4. Navigate to the new scope, click on the `Mappers` tab, and click on the `Add mapper` button.
> 5. Select `Attribute` and click on next.
> 6. Fill in the form like this and click on the Save button to confirm the creation.
> 
> Parameter	Value
> name	test
> token claim name	test
> claim json type	string
> user attribute	testclaims
> Now the scope is configured; you can pass it into the authorization request. The value of the`testclaims` claim will be passed into the identity token.

I see. Thank you for your reply. Could you please take a look at this issue: https://github.com/simpleidserver/SimpleIdServer/issues/611#issuecomment-1811715589

**Comment by simpleidserver** [comment]
@qq1176914912 : The issue #633 (`unknown scopes address`) is fixed in the `master` branch.

**Comment by qq1176914912** [comment]
> This is the expected behavior as per the OPENID RFC. When the authorization code grant type is used, the id_token does not contain the user's claims. Instead, the website utilizes the access token to retrieve the user's claims from the `userinfo` API.
> 
> Could you please verify on your website whether the claims are present in the `User.Claims` property?
> 
> I have created a bug ticket, #633, to address the issue of "unknown scopes address`.

I have roughly understood that if I want to obtain a user role and perform access control as shown in the following figure, I need to manually call the userinfo API before authorization, obtain the user role, and then manually add the role to the claims
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/8962678e-ee89-4495-8ed6-502f16709cf1)

**Comment by qq1176914912** [comment]
I have successfully implemented the functionality I need, and I customized a method before authorizing the API project:
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/bdb4a117-3d38-428a-adff-8ec23a70e684)
In this method, call userinfo to query the role, and then add the role to the claims
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/47648adb-df6f-4622-8ad8-68a1d55a45e0)
Declaration of authorization for roles on API interface, successful testing
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/77011633-a8be-4d88-9463-2192808c628a)

This is a way I came up with, I'm not sure if there are any other ways

**Comment by qq1176914912** [comment]
There is another question, what is the session in the user, and I see that they can reject it. What is the specific rejection function? If the user is already logged in and I reject it from here, the user will not redirect to the login page, so I want to know what can be done here
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/1842da10-7bd4-4178-a626-d0cb791a2044)

**Comment by simpleidserver** [comment]
@qq1176914912 

**Custom Middleware**

There is no need to implement custom middleware to retrieve claims from the userinfo endpoint, as this functionality is already implemented by `Microsost.AspNetCore.Authentication.OpenIdConnect`.
If the property `GetClaimsFromUserInfoEndpoint` is set to true, the `OpenIdConnectHandler` class retrieves the claims from the userinfo endpoint: https://github.com/dotnet/aspnetcore/blob/main/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L795

**Session**

The concept of a user's session has been introduced by the RFC `OpenID Connect Session Management (1.0)` (https://openid.net/specs/openid-connect-session-1_0.html).

In short, when the web application is configured to use `RP Initiated Logout`, and the session is rejected, the end-user will automatically be disconnected from the Client/Web Application and the Identity Server.

You can follow this tutorial to implement a SPA application with session checks enabled; the source code can be found here: https://github.com/simpleidserver/SimpleIdServer/blob/master/samples/ProjectSPA/src/Website/

**Comment by qq1176914912** [comment]
> @qq1176914912
> 
> **Custom Middleware**
> 
> There is no need to implement custom middleware to retrieve claims from the userinfo endpoint, as this functionality is already implemented by `Microsost.AspNetCore.Authentication.OpenIdConnect`. If the property `GetClaimsFromUserInfoEndpoint` is set to true, the `OpenIdConnectHandler` class retrieves the claims from the userinfo endpoint: https://github.com/dotnet/aspnetcore/blob/main/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L795
> 
> **Session**
> 
> The concept of a user's session has been introduced by the RFC `OpenID Connect Session Management (1.0)` (https://openid.net/specs/openid-connect-session-1_0.html).
> 
> In short, when the web application is configured to use `RP Initiated Logout`, and the session is rejected, the end-user will automatically be disconnected from the Client/Web Application and the Identity Server.
> 
> You can follow this tutorial to implement a SPA application with session checks enabled; the source code can be found here: https://github.com/simpleidserver/SimpleIdServer/blob/master/samples/ProjectSPA/src/Website/


Yes, what you said is correct. I checked your source code in the project 'SimpleIdServer.' I found this judgment in OpenIdConnect, but I found that it only exists in your 'SimpleIdServer.' IdServer Website The 'Startup' project references it, shouldn't this judgment be in 'SimpleIdServer.' IdServer Is it referenced in the 'Startup' project? If you follow your method, only when I log in to your 'SimpleIdServer.' IdServer Website I only judge when I start the 'Startup' project, which leads me to log in to my customized client, even if 'options.' GetClaimsFromUserInfoEndpoint=true; '

**Comment by qq1176914912** [comment]
Thanks for your hard work..I have discovered a problem.When I set BackChannelLogoutUri in the client
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/248624cc-9885-43f6-92eb-0d0110d01b7e)
When I use my client to exit, clicking on 'Revoke session' will result in an error:
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/c5ee173f-4837-4a5a-9e0c-7fb9721b8113)
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/d467b409-aa24-4e45-b181-016fc908ef76)
I also configured BackChannelLogoutUri on my IDS4 sample project, but I did not encounter this issue when I exited, so I think this issue lies in the logic of the 'Revoke session'

**Comment by simpleidserver** [comment]
Role:

By default, when using the `Microsoft.AspNetCore.Authentication.OpenIdConnect` NuGet Package, the `role` claim is not returned. 
To retrieve this claim from the userinfo endpoint, add the following code:

``` 
.AddOpenIdConnect("sid", options =>
{
	...
    options.GetClaimsFromUserInfoEndpoint = true;
    options.Scope.Add("role");
    options.ClaimActions.MapJsonKey("role", "role");
	...
});
```

Session:

You received this exception because the `TokenSignedResponseAlg` parameter is not configured in your client. 
I made some changes in the master branch to provide the ability to update this parameter in the UI. 
To do so:

* Navigate to the `Clients` screen
* Select your client and go to the `Advanced` tab.
* Set the `token signature algorithm` to `RS256` and click on `Update`.

This parameter is necessary to build and sign the `Logout Token`.

The sample project [https://github.com/simpleidserver/SimpleIdServer/tree/master/samples/ProtectWebsiteServerside](https://github.com/simpleidserver/SimpleIdServer/tree/master/samples/ProtectWebsiteServerside) 
has been updated to support Logout functionality and also to retrieve the `role` claim from the userinfo endpoint.

Feel free to explore this project! :)

**Comment by qq1176914912** [comment]
> Role:
> 
> By default, when using the `Microsoft.AspNetCore.Authentication.OpenIdConnect` NuGet Package, the `role` claim is not returned. To retrieve this claim from the userinfo endpoint, add the following code:
> 
> ```
> .AddOpenIdConnect("sid", options =>
> {
> 	...
>     options.GetClaimsFromUserInfoEndpoint = true;
>     options.Scope.Add("role");
>     options.ClaimActions.MapJsonKey("role", "role");
> 	...
> });
> ```
> 
> Session:
> 
> You received this exception because the `TokenSignedResponseAlg` parameter is not configured in your client. I made some changes in the master branch to provide the ability to update this parameter in the UI. To do so:
> 
> * Navigate to the `Clients` screen
> * Select your client and go to the `Advanced` tab.
> * Set the `token signature algorithm` to `RS256` and click on `Update`.
> 
> This parameter is necessary to build and sign the `Logout Token`.
> 
> The sample project https://github.com/simpleidserver/SimpleIdServer/tree/master/samples/ProtectWebsiteServerside has been updated to support Logout functionality and also to retrieve the `role` claim from the userinfo endpoint.
> 
> Feel free to explore this project! :)

About Exiting
1. Have you ever considered that when you click 'Reject' in 'Users' ->' Sessions', the client logged in with this account will also exit. Currently, I have found that it does not support it.
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/e034e121-a881-4bd9-bea5-31079d7691de)
2. Regarding the 'Backchannel Logout URL' configured by the client, the normal logic should be that I have configured this address. On the same browser, when logging into my client and accessing the address of the ids (5001 for your project), there is no need to log in again. When I log out of this account on this ids, he should send a message to the 'Backchannel Logout URL' address I have configured, And this address can be the Signout for the project you provided, so that I can ensure that in the same browser, if I log out of this account on IDS, my colleagues who log in to this account will also log out of the client. Unfortunately, when I tried to test this feature with your project, I found that he did not send a message to the 'Backchannel Logout URL' address I configured

**Comment by simpleidserver** [comment]
Firstly, thank you for your feedback; it is very valuable for us to enhance our product :)

The implementation is currently missing, and I am actively working on it in the `Ticket624` branch

**Comment by qq1176914912** [comment]
> Firstly, thank you for your feedback; it is very valuable for us to enhance our product :)
> 
> The implementation is currently missing, and I am actively working on it in the `Ticket624` branch

So, regarding the issue of 'Backchannel Logout URL' not working, is it also not implemented on your end?

**Comment by qq1176914912** [comment]
There is another issue, I couldn't find a place to configure the AccessionTokenType on the client side. It has two types of AccessionTokenTypes Jwt, and AccessiTokenType Reference

**Comment by simpleidserver** [comment]
The `BackChannel Logout URL` has already been implemented in SimpleIdServer.

However, I noticed two issues:

* When the session is rejected from the Administration Website, the logic used to end the session is not executed. Nonetheless, this logic has already been implemented and can be found here: https://github.com/simpleidserver/SimpleIdServer/blob/0b66bb4cd54190608eadb5c7673f0c0f5ee928d7/src/IdServer/SimpleIdServer.IdServer/UI/CheckSessionController.cs#L92.
* When the session is rejected, only the first client using the session is notified, and not all clients.

Once the changes are merged into the master branch, I will send you a sample project.

Currently, SimpleIdServer only supports access tokens in JWT format. 
Other formats, such as Reference, are not yet supported because they are not standard attributes in OAuth 2.0. I have created a new ticket, #640, to address the support for the Reference format.

**Comment by qq1176914912** [comment]
> By default, when using the `Microsoft.AspNetCore.Authentication.OpenIdConnect` NuGet Package, the `role` claim is not returned. To retrieve this claim from the userinfo endpoint, add the following code:

Role:
Following your method, I successfully displayed the role on the client front-end:
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/de315d6e-71ec-4fe5-a7af-89dad2a4550e)
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/ece104ec-46d8-4fd7-a9b6-f5bd8a887278)
But the problem is that whether it is the accesstoken or the idtoken, the role is still not included. What I need is to make the accesstoken include the role.

**Comment by qq1176914912** [comment]
> The `BackChannel Logout URL` has already been implemented in SimpleIdServer.
> 
> However, I noticed two issues:
> 
> * When the session is rejected from the Administration Website, the logic used to end the session is not executed. Nonetheless, this logic has already been implemented and can be found here: https://github.com/simpleidserver/SimpleIdServer/blob/0b66bb4cd54190608eadb5c7673f0c0f5ee928d7/src/IdServer/SimpleIdServer.IdServer/UI/CheckSessionController.cs#L92
>   .
> * When the session is rejected, only the first client using the session is notified, and not all clients.
> 
> Once the changes are merged into the master branch, I will send you a sample project.
> 
> Currently, SimpleIdServer only supports access tokens in JWT format. Other formats, such as Reference, are not yet supported because they are not standard attributes in OAuth 2.0. I have created a new ticket, #640, to address the support for the Reference format.

The problem is not just with session rejection here, you can try it out. Normally, if the 'Backchannel Logout URL' is configured, if you log in to the client in the same browser and access ids (5001) in the same browser, and exit the account on ids, It should also trigger a request to this address (I referred to Identity Server 4 and configured the Backchannel Logout URL using its ids and tested it), but in actual use, your project did not trigger it.

**Comment by simpleidserver** [comment]
@Role :

According to the OPENID specification, it is normal for the `id_token` not to contain the requested claims.
 Claims are returned in the `id_token` when the response_type is equal to `id_token`.

Test Plan : https://www.certification.openid.net/log-detail.html?log=4wDrNyZLDgxFUKg&public=true

URL : https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims

```
When a response_type value is used that results in an Access Token being issued. However, when no Access Token is issued (which is the case for the response_type value id_token), the resulting Claims are returned in the ID Token.
```

At the moment, roles are only returned in the `id_token` because there is no requirement from the OPENID standard to return them into the access token. 
However, I can understand this requirement, and I have created a new ticket #641 to add the mapping into the `API Resource`.

@Session :

It is working on my local machine; can you please follow these steps:

1. Get the latest changes in the master branch.
2. Run the administration website and add a new client.
3. On the Clients, screen, click on the Add client button.
4. Select `web application` and click on next.
5. Fill-in the form link this and click on the `Save` button to confirm the creation.

| Parameter | Value |
| ---------- | ---------------------- |
| Identifier | protectedServersideApp |
| Secret     | password               |
| Name       | protectedServersideApp |
| Redirection URLs | http://localhost:7000/signin-oidc |

6. Navigate to the new client and update the following properties :

| Parameter | Value |
| ---------- | ------- |
| Back Channel Logout session required |  Checked |
| Backchannel logout url | http://localhost:7000/BackChannelLogout/Logout |
| Validation post logout redirect URIs | http://localhost:7000/signout-callback-oidc |

7. Assign the `role` scope.
8. Navigate to the sample project `samples\ProtectWebsiteServerside\src\Website\` and run it.
9. Browse the URL `http://localhost:7000/claims` and authenticate with your credentials.
10. Click on the `Logout` button and click on `Revoke` button. The endpoint http://localhost:7000/BackChannelLogout/Logout will be called by SimpleIdServer.

I also tested with the sample project from `Duende` (successor of IdServer), and it is working : https://github.com/DuendeSoftware/Samples/tree/main/IdentityServer/v7/SessionManagement/BackChannelClient

**Comment by qq1176914912** [comment]
> @ROLE :
> 
> According to the OPENID specification, it is normal for the `id_token` not to contain the requested claims. Claims are returned in the `id_token` when the response_type is equal to `id_token`.
> 
> Test Plan : https://www.certification.openid.net/log-detail.html?log=4wDrNyZLDgxFUKg&public=true
> 
> URL : https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims
> 
> ```
> When a response_type value is used that results in an Access Token being issued. However, when no Access Token is issued (which is the case for the response_type value id_token), the resulting Claims are returned in the ID Token.
> ```
> 
> At the moment, roles are only returned in the `id_token` because there is no requirement from the OPENID standard to return them into the access token. However, I can understand this requirement, and I have created a new ticket #641 to add the mapping into the `API Resource`.
> 
> @Session :
> 
> It is working on my local machine; can you please follow these steps:
> 
> 1. Get the latest changes in the master branch.
> 2. Run the administration website and add a new client.
> 3. On the Clients, screen, click on the Add client button.
> 4. Select `web application` and click on next.
> 5. Fill-in the form link this and click on the `Save` button to confirm the creation.
> 
> Parameter	Value
> Identifier	protectedServersideApp
> Secret	password
> Name	protectedServersideApp
> Redirection URLs	http://localhost:7000/signin-oidc
> 6. Navigate to the new client and update the following properties :
> 
> Parameter	Value
> Back Channel Logout session required	Checked
> Backchannel logout url	http://localhost:7000/BackChannelLogout/Logout
> Validation post logout redirect URIs	http://localhost:7000/signout-callback-oidc
> 7. Assign the `role` scope.
> 8. Navigate to the sample project `samples\ProtectWebsiteServerside\src\Website\` and run it.
> 9. Browse the URL `http://localhost:7000/claims` and authenticate with your credentials.
> 10. Click on the `Logout` button and click on `Revoke` button. The endpoint http://localhost:7000/BackChannelLogout/Logout will be called by SimpleIdServer.
> 
> I also tested with the sample project from `Duende` (successor of IdServer), and it is working : https://github.com/DuendeSoftware/Samples/tree/main/IdentityServer/v7/SessionManagement/BackChannelClient


ROLE :
Yes, I need the accesstoken to be able to carry a role, because the client accesses the API with the accesstoken (jwt). It seems that I can only use the custom middleware mentioned earlier to add a role to the claims through userinfo.
Session：
You are right. Currently, your program exits from the client and clicking 'Revoke' can trigger the set Backchannel Logout URL normally, but you have not attempted this operation:

1. Log in to the client in the browser

2. After the client login is completed, open a new page in the current browser to access the 5001/master page

3. Exit the current user on 5001/master

In this situation, you will not trigger the 'Backchannel Logout URL'

You can try what you said‘ https://github.com/DuendeSoftware/Samples/tree/main/IdentityServer/v7/SessionManagement/BackChannelClient ’The project can still trigger the 'Backchannel Logout URL' when exiting the user on 5001, rather than having to click on 'Logout' on the client to trigger it.
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/ba4ea787-3613-43a9-9f6f-7d2354787694)
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/1dd986be-59f1-4fb0-bee0-7f089ae555b7)
![image](https://github.com/simpleidserver/SimpleIdServer/assets/79817742/f83bba57-c732-449e-9d6b-c50bfc135aa9)

**Comment by simpleidserver** [comment]
My apologies, the session was not ending when the end-user clicked on the 'Logout' button on the IdServer website (https://localhost:5001/master). 
This issue has now been resolved in the 'master' branch.

Additionally, I have created tickets for the following issues:

* Ticket #641: Add role to access token.
* Ticket #640: Support access token with reference format.

You can track their progress on the project board:
https://github.com/users/simpleidserver/projects/6/views/1

**Comment by qq1176914912** [comment]
> My apologies, the session was not ending when the end-user clicked on the 'Logout' button on the IdServer website (https://localhost:5001/master). This issue has now been resolved in the 'master' branch.
> 
> Additionally, I have created tickets for the following issues:
> 
> * Ticket [IdServer - Add custom claim in Api Resource (Role / SessionId) #641](https://github.com/simpleidserver/SimpleIdServer/issues/641): Add role to access token.
> * Ticket [IdServer - Supports AccessTokenType (JWT or Reference) #640](https://github.com/simpleidserver/SimpleIdServer/issues/640): Support access token with reference format.
> 
> You can track their progress on the project board: https://github.com/users/simpleidserver/projects/6/views/1

Thank you very much for your patient answer during this period.
